lib/storage/metricnamestats/tracker.go

```diff
@@ -0,0 +1,572 @@
+	if err != nil {
+		return fmt.Errorf("cannot open file for state save: %w", err)
+	}
+	defer f.Close()
+
+	zw := gzip.NewWriter(f)
+	writer := json.NewEncoder(zw)
+	if err := writer.Encode(mt.maxSizeBytes); err != nil {
+		return fmt.Errorf("cannot save encoded maxSizeBytes: %w", err)
+	}
+	if err := writer.Encode(mt.creationTs.Load()); err != nil {
+		return fmt.Errorf("cannot save encoded creation timestamp: %w", err)
+	}
+
+	var r recordForStore
+	for sk, si := range mt.store {
+		r.AccountID = sk.accountID
+		r.ProjectID = sk.projectID
+		r.MetricName = sk.metricName
+		r.LastRequestTs = si.lastRequestTs.Load()
+		r.RequestsCount = si.requestsCount.Load()
+		if err := writer.Encode(r); err != nil {
+			return fmt.Errorf("cannot save encoded state record: %w", err)
+		}
+	}
+	if err := zw.Close(); err != nil {
+		return fmt.Errorf("cannot flush writer state: %w", err)
+	}
+	// atomically save result
+	if err := os.Rename(f.Name(), mt.cachePath); err != nil {
+		return fmt.Errorf("cannot move temporary file %q to %q: %s", f.Name(), mt.cachePath, err)
+	}
+	return nil
+}
+
+// TrackerMetrics holds metrics to report
+type TrackerMetrics struct {
+	CurrentSizeBytes  uint64
+	CurrentItemsCount uint64
+	MaxSizeBytes      uint64
+}
+
+// UpdateMetrics writes internal metrics to the provided object
+func (mt *Tracker) UpdateMetrics(dst *TrackerMetrics) {
+	if mt == nil {
+		return
+	}
+	dst.CurrentSizeBytes = mt.currentSizeBytes.Load()
+	dst.CurrentItemsCount = mt.currentItemsCount.Load()
+	dst.MaxSizeBytes = mt.maxSizeBytes
+}
+
+// IsEmpty checks if internal state has any records
+func (mt *Tracker) IsEmpty() bool {
+	return mt.currentItemsCount.Load() > 0
+}
+
+// Reset cleans stats, saves cache state and executes provided func
+func (mt *Tracker) Reset(onReset func()) {
+	if mt == nil {
+		return
+	}
+	logger.Infof("reseting metric names tracker state")
+	mt.mu.Lock()
+	defer mt.mu.Unlock()
+	mt.store = make(map[statKey]*statItem)
+	mt.metricNamesBuf = make([]byte, 0, 16*1024)
+	mt.statItemBuf = make([]statItem, 0, 1024)
+	mt.currentSizeBytes.Store(0)
+	mt.currentItemsCount.Store(0)
+	mt.creationTs.Store(mt.getCurrentTs())
+	if err := mt.saveLocked(); err != nil {
+		logger.Panicf("during Tracker reset cannot save state: %s", err)
+	}
+	onReset()
+}
+
+// RegisterIngestRequest tracks metric name ingestion
+func (mt *Tracker) RegisterIngestRequest(accountID, projectID uint32, metricGroup []byte) {
+	if mt == nil {
+		return
+	}
+	if mt.shouldSkipNewItemAdd() {
+		return
+	}
+
+	sk := statKey{accountID: accountID, projectID: projectID, metricName: bytesutil.ToUnsafeString(metricGroup)}
+	mt.mu.RLock()
+	_, ok := mt.store[sk]
+	mt.mu.RUnlock()
+	if ok {
+		return
+	}
+
+	mt.mu.Lock()
+	// key could be already ingested concurrently
+	_, ok = mt.store[sk]
+	if ok {
+		mt.mu.Unlock()
+		return
+	}
+	si := mt.nextRecordLocked()
+	sk.metricName = bytesutil.ToUnsafeString(mt.allocateMetricNameLocked(metricGroup))
```

Comment:
I don't understand how it is save to do the unsafe conversion here... On the next iteration `mt.metricNamesBuf` can be re-allocated and `sk.metricName` will start showing trash.

Reply:
I think, it's better to rename this function into `mt.cloneMetricNameLocked`. This code is totally safe.